types/tablet_v2/wlr_tablet_v2_pad.c

```diff
@@ -0,0 +1,564 @@
+	struct wlr_tablet_seat_v2 *seat = get_or_create_tablet_seat(manager, wlr_seat);
+	if (!seat) {
+		return NULL;
+	}
+	struct wlr_tablet_pad *wlr_pad = wlr_device->tablet_pad;
+	struct wlr_tablet_v2_tablet_pad *pad = calloc(1, sizeof(struct wlr_tablet_v2_tablet_pad));
+	if (!pad) {
+		return NULL;
+	}
+
+	pad->group_count = wl_list_length(&wlr_pad->groups);
+	pad->groups = calloc(pad->group_count, sizeof(int));
+	if (!pad->groups) {
+		free(pad);
+		return NULL;
+	}
+
+	pad->wlr_pad = wlr_pad;
+	wl_list_init(&pad->clients);
+
+	pad->pad_destroy.notify = handle_wlr_tablet_pad_destroy;
+	wl_signal_add(&wlr_device->events.destroy, &pad->pad_destroy);
+	wl_list_insert(&seat->pads, &pad->link);
+
+	// We need to create a tablet client for all clients on the seat
+	struct wlr_tablet_seat_client_v2 *pos;
+	wl_list_for_each(pos, &seat->clients, seat_link) {
+		// Tell the clients about the new tool
+		add_tablet_pad_client(pos, pad);
+	}
+
+	wl_signal_init(&pad->events.button_feedback);
+	wl_signal_init(&pad->events.strip_feedback);
+	wl_signal_init(&pad->events.ring_feedback);
+
+	return pad;
+}
+
+struct wlr_tablet_pad_client_v2 *tablet_pad_client_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_tablet_pad_v2_interface,
+		&tablet_pad_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+/* Actual protocol foo */
+uint32_t wlr_send_tablet_v2_tablet_pad_enter(
+		struct wlr_tablet_v2_tablet_pad *pad,
+		struct wlr_tablet_v2_tablet *tablet,
+		struct wlr_surface *surface) {
+	struct wl_client *client = wl_resource_get_client(surface->resource);
+
+	struct wlr_tablet_client_v2 *tablet_tmp;
+	struct wlr_tablet_client_v2 *tablet_client = NULL;
+	wl_list_for_each(tablet_tmp, &tablet->clients, tablet_link) {
+		if (tablet_tmp->client == client) {
+			tablet_client = tablet_tmp;
+			break;
+		}
+	}
+
+	// Couldn't find the client binding for the surface's client. Either
+	// the client didn't bind tablet_v2 at all, or not for the relevant
+	// seat
+	if (!tablet_client) {
+		return 0;
+	}
+
+	struct wlr_tablet_pad_client_v2 *pad_tmp = NULL;
+	struct wlr_tablet_pad_client_v2 *pad_client = NULL;
+	wl_list_for_each(pad_tmp, &pad->clients, pad_link) {
+		if (pad_tmp->client == client) {
+			pad_client = pad_tmp;
+			break;
+		}
+	}
+
+	// Couldn't find the client binding for the surface's client. Either
+	// the client didn't bind tablet_v2 at all, or not for the relevant
+	// seat
+	if (!pad_client) {
+		return 0;
+	}
+
+	pad->current_client = pad_client;
+
+	/* Pre-increment keeps 0 clean. wraparound would be after 2^32
+	 * proximity_in. Someone wants to do the math how long that would take?
+	 */
+	uint32_t serial = wl_display_next_serial(wl_client_get_display(client));
+
+	zwp_tablet_pad_v2_send_enter(pad_client->resource, serial,
+		tablet_client->resource, surface->resource);
+
+	struct timespec now;
+	clock_gettime(CLOCK_MONOTONIC, &now);
+	uint32_t time = now.tv_nsec / 1000;
```

Comment:
Can't we get the timestamp from libinput?

Reply:
Currently this is bound to the seat keyboard focus. If that's changed, or we require the time to be passed to that as well then probably? But focus change there can also happen from a client unmap, which doesn't have the timestamp from libinput.

Reply:
Yeah, probably better to keep it that way.